File: src/main/java/um/tesoreria/report/hexagonal/chequeras/domain/model/ChequeraSerie.java
package um.tesoreria.report.hexagonal.chequeras.domain.model;

import lombok.*;
import um.tesoreria.report.util.Jsonifier;
import java.math.BigDecimal;
import java.time.OffsetDateTime;

@Getter
@Setter
@Builder
@NoArgsConstructor
@AllArgsConstructor
public class ChequeraSerie {
    private Long chequeraId;
    private Integer facultadId;
    private Integer tipoChequeraId;
    private Long chequeraSerieId;
    private BigDecimal personaId;
    private Integer documentoId;
    private Integer lectivoId;
    private Integer arancelTipoId;
    private Integer cursoId;
    private Byte asentado;
    private Integer geograficaId;
    private OffsetDateTime fecha;
    private Integer cuotasPagadas;
    private String observaciones;
    private Integer alternativaId;
    private Byte algoPagado;
    private Integer tipoImpresionId;
    private Byte flagPayperTic;
    private String usuarioId;
    private Byte enviado;
    private Byte retenida;
    private Long version;
    private Byte hpum;
    private BigDecimal becaPorcentaje;
    private String becaResolucion;
    private OffsetDateTime becaFecha;
    private Long becaUserId;
    private Integer cuotasDeuda;
    private BigDecimal importeDeuda;
    private OffsetDateTime ultimoEnvio;

    private Facultad facultad;
    private TipoChequera tipoChequera;
    private Persona persona;
    private Domicilio domicilio;
    private Lectivo lectivo;
    private ArancelTipo arancelTipo;
    private Geografica geografica;

    public String jsonify() {
        return Jsonifier.builder(this).build();
    }
}

File: src/main/java/um/tesoreria/factura/rest/kotlin/model/ChequeraSerie.kt
package um.tesoreria.factura.rest.kotlin.model

import um.tesoreria.factura.rest.model.Auditable
import com.fasterxml.jackson.annotation.JsonFormat
import jakarta.persistence.*
import java.math.BigDecimal
import java.time.OffsetDateTime

@Entity
@Table(uniqueConstraints = [UniqueConstraint(columnNames = ["chs_fac_id", "chs_tch_id", "chs_id"])])
data class ChequeraSerie(

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    @Column(name = "clave")
    var chequeraId: Long? = null,

    @Column(name = "chs_fac_id")
    var facultadId: Int? = null,

    @Column(name = "chs_tch_id")
    var tipoChequeraId: Int? = null,

    @Column(name = "chs_id")
    var chequeraSerieId: Long? = null,

    @Column(name = "chs_per_id")
    var personaId: BigDecimal? = null,

    @Column(name = "chs_doc_id")
    var documentoId: Int? = null,

    @Column(name = "chs_lec_id")
    var lectivoId: Int? = null,

    @Column(name = "chs_art_id")
    var arancelTipoId: Int? = null,

    @Column(name = "chs_cur_id")
    var cursoId: Int? = null,

    @Column(name = "chs_asentado")
    var asentado: Byte? = null,

    @Column(name = "chs_geo_id")
    var geograficaId: Int? = null,

    @Column(name = "chs_fecha")
    @JsonFormat(shape = JsonFormat.Shape.STRING, pattern = "yyyy-MM-dd'T'HH:mm:ssZ", timezone = "UTC")
    var fecha: OffsetDateTime? = null,

    @Column(name = "chs_cuotasp")
    var cuotasPagadas: Int? = null,

    @Column(name = "chs_observ")
    var observaciones: String? = null,

    @Column(name = "chs_alt_id")
    var alternativaId: Int? = null,

    @Column(name = "chs_algopagado")
    var algoPagado: Byte? = null,

    @Column(name = "chs_tim_id")
    var tipoImpresionId: Int? = null,

    @Column(name = "flag_paypertic")
    var flagPayperTic: Byte = 0,

    var usuarioId: String? = null,
    var enviado: Byte = 0,
    var retenida: Byte = 0,

    @Transient
    var cuotasDeuda: Int = 0,

    @Transient
    var importeDeuda: BigDecimal = BigDecimal.ZERO,

    @OneToOne(optional = true)
    @JoinColumn(name = "chs_fac_id", insertable = false, updatable = false)
    var facultad: Facultad? = null,

    @OneToOne(optional = true)
    @JoinColumn(name = "chs_tch_id", insertable = false, updatable = false)
    var tipoChequera: TipoChequera? = null,

    @OneToOne(optional = true)
    @JoinColumns(
        JoinColumn(name = "chs_per_id", referencedColumnName = "per_id", insertable = false, updatable = false),
        JoinColumn(name = "chs_doc_id", referencedColumnName = "per_doc_id", insertable = false, updatable = false)
    )
    var persona: Persona? = null,

    @OneToOne(optional = true)
    @JoinColumns(
        JoinColumn(name = "chs_per_id", referencedColumnName = "dom_per_id", insertable = false, updatable = false),
        JoinColumn(name = "chs_doc_id", referencedColumnName = "dom_doc_id", insertable = false, updatable = false),
    )
    var domicilio: Domicilio? = null,

    @OneToOne(optional = true)
    @JoinColumn(name = "chs_lec_id", insertable = false, updatable = false)
    var lectivo: Lectivo? = null,

    @OneToOne(optional = true)
    @JoinColumn(name = "chs_art_id", insertable = false, updatable = false)
    var arancelTipo: ArancelTipo? = null,

    @OneToOne(optional = true)
    @JoinColumn(name = "chs_geo_id", insertable = false, updatable = false)
    var geografica: Geografica? = null,

    ) : Auditable() {
    fun getPersonaKey(): String {
        return this.personaId.toString() + "." + this.documentoId
    }

    fun getFacultadKey(): String {
        return this.facultadId.toString() + "." + this.lectivoId + "." + this.geograficaId + "." + this.getPersonaKey()
    }

}
